### ISSUE TITLE
Does spack work on alpine?

### ISSUE THREAD

**Issue by vsoch** [spack/spack#11370]
I've installed g++, gfortran, and all the build tools I'm able to find for alpine, but I keep getting:

```
2 errors found in build log:
  >> 6    configure: error: in `/tmp/root/spack-stage/spack-stage-_73gNo/pkgconf-1.6.0':
  >> 7    configure: error: C compiler cannot create executables
     8    See `config.log' for more details

See build log for details:
  /home/spack/var/spack/stage/pkgconf-1.6.0-5k3ytveiqoecv6huh6fulebbotrhcgy4/spack-build.out
/ # cat /home/spack/var/spack/stage/pkgconf-1.6.0-5k3ytveiqoecv6huh6fulebbotrhcgy4/spack-build.out
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> [2019-05-04-15:47:42.524938] '/home/spack/var/spack/stage/pkgconf-1.6.0-5k3ytveiqoecv6huh6fulebbotrhcgy4/pkgconf-1.6.0/configure' '--prefix=/home/spack/opt/spack/linux-alpine3-x86_64/gcc-8.3.0/pkgconf-1.6.0-5k3ytveiqoecv6huh6fulebbotrhcgy4'
checking for gcc... /home/spack/lib/spack/env/gcc/gcc
checking whether the C compiler works... no
configure: error: in `/tmp/root/spack-stage/spack-stage-_73gNo/pkgconf-1.6.0':
configure: error: C compiler cannot create executables
See `config.log' for more details
```
Has anyone tried this before? I thought it would be a fairly easy fix, but I'm going around now installing random packages and can't get it working.

**Comment by vsoch** [comment]
Wooooo just got it working! Nevermind :)

**Comment by vsoch** [comment]
Here's a recipe for all the future turkeys that feel like building in alpine :)

```
Bootstrap: docker
From: alpine

%environment
SPACK_ROOT=/home/spack
FORCE_UNSAFE_CONFIGURE=1
PATH="${PATH}:${SPACK_ROOT}/bin"
export SPACK_ROOT FORCE_UNSAFE_CONFIGURE PATH

%help
This is a spack container with packages curl

%post
SPACK_ROOT=/home/spack
FORCE_UNSAFE_CONFIGURE=1
PATH="${PATH}:${SPACK_ROOT}/bin"
export SPACK_ROOT FORCE_UNSAFE_CONFIGURE PATH

# Install basic system dependencies
apk update && \
    apk add --no-cache git gcc g++ gfortran make bzip2 && \
    apk add --no-cache patch file curl python gnupg xz && \
    apk add --no-cache curl bash openssh libtool
    
# Install spack + set the mirror
mkdir -p $SPACK_ROOT
curl -s -L https://api.github.com/repos/spack/spack/tarball/develop \
    | tar xzC $SPACK_ROOT --strip 1

/home/spack/bin/spack install curl

%runscript
exec /home/spack/bin/spack "$@"
```
Turkey power! :turkey:

**Comment by Sinan81** [comment]
Thanks! I was just looking for this. Can this be mentioned in Spack docs along side ubuntu/centos ? It seems Alpine is very popular.

**Comment by Sinan81** [comment]
How do we build the above mentioned spack alpine image? Can you provided specific instructions. I am not super familiar with Docker.

**Comment by Sinan81** [comment]
I just combined the above mentioned steps with the recipe in Spack centos7 dockerfile:
``` dockerfile
FROM alpine:3.10
MAINTAINER Spack Maintainers <maintainers@spack.io>

ENV DOCKERFILE_BASE=centos            \
    DOCKERFILE_DISTRO=centos          \
    DOCKERFILE_DISTRO_VERSION=7       \
    SPACK_ROOT=/opt/spack             \
    DEBIAN_FRONTEND=noninteractive    \
    CURRENTLY_BUILDING_DOCKER_IMAGE=1 \
    container=docker

COPY bin   $SPACK_ROOT/bin
COPY etc   $SPACK_ROOT/etc
COPY lib   $SPACK_ROOT/lib
COPY share $SPACK_ROOT/share
COPY var   $SPACK_ROOT/var
RUN mkdir -p $SPACK_ROOT/opt/spack

RUN ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/docker-shell \
 && ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/interactive-shell \
 && ln -s $SPACK_ROOT/share/spack/docker/entrypoint.bash \
          /usr/local/bin/spack-env

# Install basic system dependencies
RUN apk update && \
    apk add --no-cache git gcc g++ gfortran make bzip2 && \
    apk add --no-cache patch file curl python gnupg xz && \
    apk add --no-cache curl bash openssh libtool linux-headers
#   pip install boto3 # amazon stuff


RUN mkdir -p /root/.spack \
 && cp $SPACK_ROOT/share/spack/docker/modules.yaml \
        /root/.spack/modules.yaml \
 && rm -rf /root/*.* /run/nologin $SPACK_ROOT/.git

# [WORKAROUND]
# https://superuser.com/questions/1241548/
#     xubuntu-16-04-ttyname-failed-inappropriate-ioctl-for-device#1253889
RUN [ -f ~/.profile ]                                               \
 && sed -i 's/mesg n/( tty -s \\&\\& mesg n || true )/g' ~/.profile \
 || true

WORKDIR /root
SHELL ["docker-shell"]

# TODO: add a command to Spack that (re)creates the package cache
RUN spack spec hdf5+mpi

ENTRYPOINT ["/bin/bash", "/opt/spack/share/spack/docker/entrypoint.bash"]
CMD ["interactive-shell"]
```

In order to build it one needs to put the Docker file in a fresh clone of spack repo:
```bash
git clone git clone https://github.com/spack/spack.git
cp Dockerfile ./spack
cd ./spack
sudo docker build --tag test_spack_alpine_image .
```
Aterwards one can enter interactive shell via
```bash
sudo docker run -it test_spack_alpine_image
```

**Comment by vsoch** [comment]
hey @Sinan81 ! I was just testing this for you, but it seems that your computer is faster than mine. The above recipe is actually for a  [Singularity](https://sylabs.io/guides/3.5/user-guide/build_a_container.html#creating-writable-sandbox-directories) container, meaning that we get layers from Docker and dump them into a root folder and build into a sif binary. To do that:

```bash
sudo singularity build spack.sif Singularity
```
where `Singularity` is in reference to the file above. If you want to build the "same" container with [Docker](https://docs.docker.com/get-started/) then here is a modified recipe (this is a file named `Dockerfile`)

```
FROM alpine

# docker build -t spack .

ENV SPACK_ROOT=/home/spack
ENV FORCE_UNSAFE_CONFIGURE=1
ENV PATH="${PATH}:${SPACK_ROOT}/bin"

# Install basic system dependencies
RUN apk update && \
    apk add --no-cache git gcc g++ gfortran make bzip2 && \
    apk add --no-cache patch file curl python3 gnupg xz && \
    apk add --no-cache curl bash openssh libtool linux-headers
    
# Install spack + set the mirror
RUN mkdir -p $SPACK_ROOT && \
    curl -s -L https://api.github.com/repos/spack/spack/tarball/develop \
    | tar xzC $SPACK_ROOT --strip 1 && \
    /home/spack/bin/spack install curl

ENTRYPOINT ["/home/spack/bin/spack"]
```
and you would do:

```bash
docker build -t spack .
```
in the root of the folder with the Dockerfile. Note that I had to update `python` to be `python3` so hopefully that doesn't introduce any bugs, if it does it's likely just needed to do a `python` alias. Also

**Comment by Sinan81** [comment]
@vsoch Thanks a lot! Appreciate it. Your version of Dockerfile looks simpler, which is great. As you guessed, I had an issue with python, for which quick fix was to go back to alpine:3.10 . Thanks for suggesting the proper way to fix it.
